File: frontend/src/types/report.ts
export type ReportReason =
    | 'SPAM'
    | 'OFFENSIVE_LANGUAGE'
    | 'HATE_SPEECH'
    | 'FAKE_PROFILE'
    | 'MISINFORMATION'
    | 'SPOILER'
    | 'OTHER';

export interface ReportCreateDTO {
    reason: ReportReason;
    reportedUserId?: string;
    reportedCommentId?: string;
    reportedTopicId?: string;
    reportedPostId?: string;
}

interface UserSummary {
    id: string;
    name: string;
    email: string;
    isCommentBanned?: boolean;
    commentBanExpiresAt?: string;
}

export interface ReportDTO {
    id: string;
    reason: string;
    status: string;
    createdAt: string;

    reporter?: UserSummary;

    reportedUser?: UserSummary;

    reportedComment?: {
        id: string;
        content: string;
        user: UserSummary;
    };

    reportedTopic?: {
        id: string;
        title: string;
        author: UserSummary;
    };

    reportedPost?: {
        id: string;
        content: string;
        author: UserSummary;
    };
}

File: backend/src/main/java/com/biblioteca/backend/entity/Report.java
package com.biblioteca.backend.entity;

import com.biblioteca.backend.enums.ReportReason;
import com.biblioteca.backend.enums.ReportStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "tb_reports")
@Getter
@Setter
public class Report {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Enumerated(EnumType.STRING)
    private ReportReason reason;

    @Enumerated(EnumType.STRING)
    private ReportStatus status = ReportStatus.PENDING;

    @Column(nullable = false)
    private Instant createdAt = Instant.now();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id", nullable = true)
    private User reporter;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reported_comment_id")
    private Comment reportedComment;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reported_user_id")
    private User reportedUser;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "topic_id")
    private ForumTopic reportedTopic;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id")
    private ForumPost reportedPost;
}